[rpc] Switch RRef to be managed by intrusive_ptr

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33263 [rpc] allow RRef local creation with IValue objects
* #33190 [jit] Initial use RRef in TorchScript
* #32992 [jit] Add RRef to IValue and JIT type system
* **#33189 [rpc] Switch RRef to be managed by intrusive_ptr**

Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

Differential Revision: [D19871241](https://our.internmc.facebook.com/intern/diff/D19871241)

wanchaol committed:
[rpc] Switch RRef to be managed by intrusive_ptr

Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Switch RRef to be managed by intrusive_ptr"


Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -263,6 +260,15 @@ class TORCH_API OwnerRRef final : public RRef {
   OwnerRRef& operator=(const OwnerRRef& other) = delete;
   OwnerRRef& operator=(OwnerRRef&& other) = delete;
 
+  OwnerRRef(worker_id_t ownerId, const RRefId& rrefId, TypePtr type)
```
IIUC, this is to make sure that we can do sth like:

```c++
c10::make_intrusive<OwnerRRef>(getWorkerId(), rrefId, type);
```

The concern here is that we expose `OwnerRRef` as `TORCH_API`, so that C++ applications can create `OwnerRRef` on their own, which can mess up with the `RRefContext`. Is it possible to avoid exposing this by passing a pointer to `intrusive_ptr<OwnerRRef>` construction API?

wanchaol replied:
tried `intrusive_ptr<OwnerRRef>(new OwnerRRef())` way, and it does not seem to work because of intrusive_ptr limitation. I will create a issue to track this.

wanchaol replied:
created https://github.com/pytorch/pytorch/issues/33246

mrshenli reviewed (commented):
LGTM! Thanks for the PR! I have a little reservation on exposing `OwnerRRef` constructor. It might be better if we could avoid that. But if this is a hard requirement, I am also OK with merging this PR as is.

wanchaol commented on aten/src/ATen/core/rref_interface.h:
```diff
@@ -1,14 +1,16 @@
+namespace c10 {
+
+struct Type;
+using TypePtr = std::shared_ptr<Type>;
```
`TypePtr` is a shared_ptr, we didn't make it an intrusive ptr(although we could), this is just for simplicity. most of the uses of intrusive_ptr is for IValue and TensorImpl, I think the current philosophy is: anything managed by IValue should use intrusive_ptr, for other pointers in codebase, if we see perf cost on shared_ptr, we should switch it.

mrshenli reviewed (approved):
Discussed with @wanchaol offline. We are going to expose `OwnerRRef` constor for now, as `intrusive_ptr` constor cannot take a raw pointer. This should be fine for now as we don't advertise the C++ API yet. But, `OwnerRRef` constor needs to be private before we are ready to announce C++ support.

@wanchaol could you link the issue to track constructing `intrusive_ptr` with raw ptr? Thanks!

wanchaol committed:
Update on "[rpc] Switch RRef to be managed by intrusive_ptr"


Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Switch RRef to be managed by intrusive_ptr"


Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Switch RRef to be managed by intrusive_ptr"


Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

Differential Revision: [D19871241](https://our.internmc.facebook.com/intern/diff/D19871241)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Switch RRef to be managed by intrusive_ptr"


Add RRefInterface to Aten/Core, which will later be used by IValue

Switch all the rpc code base to use intrusive_ptr instead of shared_ptr,
so that we could add it to IValue.

Actual adding to IValue and JIT will be in next PR

Differential Revision: [D19871241](https://our.internmc.facebook.com/intern/diff/D19871241)

[ghstack-poisoned]
